tests/unit/include/bout/test_single_index_ops.cxx

```diff
@@ -0,0 +1,303 @@
+  Field3D difops = DDZ(input);
+
+  // Differentiate using index operations
+  Field3D indexops; indexops.allocate();
+  auto input_acc = FieldAccessor<>(input);
+  BOUT_FOR(i, input.getRegion("RGN_NOBNDRY")) {
+    indexops[i] = DDZ(input_acc, i);
+  }
+
+  // Check the answer is the same
+  ASSERT_TRUE(IsFieldEqual(difops, indexops, "RGN_NOBNDRY"));
+}
+
+#if !BOUT_USE_METRIC_3D
+// Arakawa bracket not implemented for 3D metrics yet
+
+TEST_F(SingleIndexOpsTest, bracket2d3d) {
+  // Fill a field with random numbers
+  std::default_random_engine re;
+
+  auto input = random_field<Field2D>(re);
+  // Check that the field is not zero
+  ASSERT_FALSE(IsFieldEqual(input, 0.0, "RGN_NOBNDRY"));
+
+  auto input2 = random_field<Field3D>(re);
+  ASSERT_FALSE(IsFieldEqual(input2, 0.0, "RGN_NOBNDRY"));
+
+  // Differentiate whole field
+  Field3D difops = bracket(input, input2, BRACKET_ARAKAWA);
+
+  // Differentiate using index operations
+  Field3D indexops; indexops.allocate();
+  auto input_acc = Field2DAccessor<>(input);
+  auto input2_acc = FieldAccessor<>(input2);
+  BOUT_FOR(i, indexops.getRegion("RGN_NOBNDRY")) {
+    indexops[i] = bracket(input_acc, input2_acc, i);
+  }
+
+  // Check the answer is the same
+  ASSERT_TRUE(IsFieldEqual(difops, indexops, "RGN_NOBNDRY"));
+}
+
+TEST_F(SingleIndexOpsTest, bracket3d3d) {
+  // Fill a field with random numbers
+  std::default_random_engine re;
+
+  auto input = random_field<Field3D>(re);
+  // Check that the field is not zero
+  ASSERT_FALSE(IsFieldEqual(input, 0.0, "RGN_NOBNDRY"));
+
+  auto input2 = random_field<Field3D>(re);
+  ASSERT_FALSE(IsFieldEqual(input2, 0.0, "RGN_NOBNDRY"));
+
+  // Differentiate whole field
+  Field3D difops = bracket(input, input2, BRACKET_ARAKAWA);
+
+  // Differentiate using index operations
+  Field3D indexops; indexops.allocate();
+  auto input_acc = FieldAccessor<>(input);
+  auto input2_acc = FieldAccessor<>(input2);
+  BOUT_FOR(i, indexops.getRegion("RGN_NOBNDRY")) {
+    indexops[i] = bracket(input_acc, input2_acc, i);
+  }
+
+  // Check the answer is the same
+  ASSERT_TRUE(IsFieldEqual(difops, indexops, "RGN_NOBNDRY"));
+}
+
+#endif // !BOUT_USE_METRIC_3D
+
+TEST_F(SingleIndexOpsTest, Delp2_3D) {
+  // Fill a field with random numbers
+  std::default_random_engine re;
+
+  auto input = random_field<Field3D>(re);
+
+  // Check that the field is not zero
+  ASSERT_FALSE(IsFieldEqual(input, 0.0, "RGN_NOBNDRY"));
+
+  // Differentiate whole field
+  // Note: Uses non-FFT version for comparison to the single index operator
+  Field3D difops = Delp2(input, CELL_DEFAULT, false);
+
+  // Differentiate using index operations
+  Field3D indexops; indexops.allocate();
+  auto input_acc = FieldAccessor<>(input);
+  BOUT_FOR(i, input.getRegion("RGN_NOBNDRY")) {
+    indexops[i] = Delp2(input_acc, i);
+  }
+
+  // Check the answer is the same
+  ASSERT_TRUE(IsFieldEqual(difops, indexops, "RGN_NOBNDRY"));
+}
+
+TEST_F(SingleIndexOpsTest, b0xGrad_dot_Grad_3D_2D) {
```

Comment:
Were there `b0xGrad_dot_Grad_2D_3D` and `b0xGrad_dot_Grad_3D_3D` variants too that could be tested?

On a related note - has anyone checked that all the single-index operators are tested here?